File: image_matcher.py
import os
import sys
import cv2
import numpy as np
import json
import time


def read_file(path):
    with open(path, "r") as file:
        return file.read().strip().split("\n")


def get_stock_data(path):
    stock_data_raw = read_file(path)
    stock_data = []
    for stock in stock_data_raw:
        stock_parts = stock.strip().split(",")
        descriptor_file_path = f"./Amrosia Stock Descriptors/{stock_parts[1]}_{stock_parts[2]}.txt"
        file_path = f"./Purse Pics/Amrosia Stock/{stock_parts[0]}/{stock_parts[1]}.jpeg"
        stock_data.append([descriptor_file_path, file_path, *stock_parts])
    return stock_data


sift = cv2.SIFT_create()


def get_descriptors(img):
    keypoints = sift.detect(img, None)
    keypoints, descriptors = sift.compute(img, keypoints)
    return descriptors


def crop_black_bars(image):
    gray_image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    gray_image = cv2.GaussianBlur(gray_image, (15, 15), 10)
    _, thresh_image = cv2.threshold(gray_image, 1, 255, cv2.THRESH_BINARY)
    contours, _ = cv2.findContours(thresh_image, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    biggest_contour = max(contours, key=cv2.contourArea)
    x, y, w, h = cv2.boundingRect(biggest_contour)
    return image[y : y + h, x : x + w]


def set_some_values_to_zero_in_hist(hist):
    hist[0][0] = 0
    hist[1][0] = 0
    hist[254][0] = 0
    hist[255][0] = 0


def find_matches_with_histogram_no_JSON(stock_array, received_img):
    hist_comp_method = cv2.HISTCMP_BHATTACHARYYA
    similarity_threshold = 0.15
    similar_stock = []
    b_hist_rec = cv2.calcHist([received_img], [0], None, [256], [0, 256])
    g_hist_rec = cv2.calcHist([received_img], [1], None, [256], [0, 256])
    r_hist_rec = cv2.calcHist([received_img], [2], None, [256], [0, 256])
    set_some_values_to_zero_in_hist(b_hist_rec)
    set_some_values_to_zero_in_hist(g_hist_rec)
    set_some_values_to_zero_in_hist(r_hist_rec)

    for index, stock in enumerate(stock_array):
        file_path = stock[1]
        tempFileName = '_'.join(file_path.split('/')[2:]).replace('.jpeg', '')
        npyFileName = f"histograms/{tempFileName}.npy"
        stored_hist = np.load(npyFileName)
        b_hist_ref = np.array(stored_hist[0], dtype=np.float32)
        g_hist_ref = np.array(stored_hist[1], dtype=np.float32)
        r_hist_ref = np.array(stored_hist[2], dtype=np.float32)
        similarity_b = cv2.compareHist(b_hist_rec, b_hist_ref, hist_comp_method)
        similarity_g = cv2.compareHist(g_hist_rec, g_hist_ref, hist_comp_method)
        similarity_r = cv2.compareHist(r_hist_rec, r_hist_ref, hist_comp_method)
        if similarity_b < similarity_threshold and similarity_g < similarity_threshold and similarity_r < similarity_threshold:
            similar_stock.append(stock)
    return similar_stock


def best_match_using_feature_detector(some_array, received_img_descriptors):
    bf = cv2.BFMatcher()
    min_error = 10000
    best_match = None

    for data in some_array:
        stored_descriptors = data[0]
        matches = bf.match(stored_descriptors, received_img_descriptors)
        matches = sorted(matches, key=lambda x: x.distance)[:100]
        error_sum = sum([match.distance for match in matches])
        if error_sum < min_error:
            min_error = error_sum
            best_match = data
    return min_error, best_match


def optimised_image_matcher(image_data_array, image):
    image = crop_black_bars(image)
    reduced_stock_data_array = find_matches_with_histogram_no_JSON(image_data_array, image)
    
    for stock in reduced_stock_data_array:
        npyFilePath = stock[0].replace('.txt', '.npy')
        if os.path.exists(npyFilePath):
            stock[0] = np.load(npyFilePath)
            continue
        with open(stock[0], "r") as file:
            stock[0] = np.array(json.load(file), dtype=np.float32)
        np.save(npyFilePath, stock[0])

    received_img_descriptors = get_descriptors(image)
    feature_detector_result = best_match_using_feature_detector(reduced_stock_data_array, received_img_descriptors)

    if feature_detector_result[0] < 9000:
        product_filepath = feature_detector_result[1][1]
        detected_image = cv2.imread(product_filepath)
        # Further processing as needed

    return feature_detector_result


def get_product_data(path):
    product_data_raw = read_file(path)
    product_data = []
    for stock in product_data_raw:
        if not stock:  # Skip empty lines
            continue
        stock_parts = stock.strip().split(",")
        descriptor_path = f"./Supplier Products/{stock_parts[4]}/{stock_parts[0]}/Descriptors/{stock_parts[1]}.txt"
        image_path = f"./Supplier Products/{stock_parts[4]}/{stock_parts[0]}/{stock_parts[1]}.jpeg"
        last_availability_update_time = json.loads(stock_parts[2])
        availability = json.loads(stock_parts[3].lower())
        price = json.loads(stock_parts[5])
        last_inquiry_update_time = json.loads(stock_parts[6])
        product_data.append([
            descriptor_path,
            image_path,
            last_availability_update_time,
            availability,
            stock_parts[4],
            price,
            last_inquiry_update_time,
        ])
    return product_data


productDataPath = "./Supplier Products/product data.csv"
productData = get_product_data(productDataPath)

image_path = sys.argv[1]  # Replace with your actual image path
image = cv2.imread(image_path, cv2.IMREAD_COLOR)

result = optimised_image_matcher(productData, image)

try:
    with open('matches.txt', 'a') as file:
        file.write(json.dumps(result) + '\n')
except Exception as e:
    print(f"An error occurred: {e}")

print(json.dumps(result))

File: automation_modules/image_matcher.js
const fs = require('fs')
const cv = require('@u4/opencv4nodejs')
const { execFile } = require('child_process');
const { promisify } = require('util');


function cvMatToArray(cvMat) {
    let array = []
    for (let i = 0; i <= cvMat.rows - 1; i++) {
        let row = []
        for (let j = 0; j <= cvMat.cols - 1; j++) {
            let point = cvMat.at(i, j)
            row.push(point)
        }
        array.push(row)
    }
    return array
}

function getStockData(path) {
    const stockDataRaw = fs.readFileSync(path, 'utf-8').trim().split('\n');

    const stockData = stockDataRaw.map(stock => {
        const stockParts = stock.trim().split(',');

        const descriptorFilePath = `./Amrosia Stock Descriptors/${stockParts[1]}_${stockParts[2]}.txt`;
        const filePath = `./Purse Pics/Amrosia Stock/${stockParts[0]}/${stockParts[1]}.jpeg`;

        return [descriptorFilePath, filePath, ...stockParts];
    });

    return stockData;
}


function getSeStockData(path) {
    const stockData = ((fs.readFileSync(path, 'utf-8'))).trim().split('\n')

    stockData.forEach((stock, index) => {
        stockData[index] = stock.trim().split(',')
    })

    stockData.forEach((stock, index) => {
        const filePath = './SE Stock/' + stock[0] + '/' + stock[1] + '.jpeg'
        stockData[index].unshift(filePath)

        const descriptorFilePath = './SE Stock Descriptors/' + stock[1] + '_' + stock[2] + '.txt'
        stockData[index].unshift(descriptorFilePath)


        stockData[index][4] = JSON.parse(stock[4])
        stockData[index][5] = JSON.parse(stock[5].toLowerCase())
        stockData[index][6] = JSON.parse(stock[6])
    })

    return stockData
}

function getProductData(path) {
    const productDataRaw = fs.readFileSync(path, 'utf-8').trim().split('\n').filter(stock => stock);
    const productData = productDataRaw.map(stock => {
        const stockParts = stock.trim().split(',');

        const descriptorPath = `./Supplier Products/${stockParts[4]}/${stockParts[0]}/Descriptors/${stockParts[1]}.txt`;
        const imagePath = `./Supplier Products/${stockParts[4]}/${stockParts[0]}/${stockParts[1]}.jpeg`;

        const descriptor = descriptorPath;
        const filePath = imagePath;
        const lastAvailabilityUpdateTime = JSON.parse(stockParts[2]);
        const availability = JSON.parse(stockParts[3].toLowerCase());
        // 4 does not need to be processed
        const price = JSON.parse(stockParts[5]);
        const lastInquiryUpdateTime = JSON.parse(stockParts[6]);

        return [descriptor, filePath, lastAvailabilityUpdateTime, availability, stockParts[4], price, lastInquiryUpdateTime];
    });

    return productData;
}


const SIFT = new cv.SIFTDetector()
const bruteForceMatcher = new cv.BFMatcher()

bHistAxis = new cv.HistAxes({
    channel: 0,
    bins: 256,
    ranges: [0, 1]
})

gHistAxis = new cv.HistAxes({
    channel: 1,
    bins: 256,
    ranges: [0, 1]
})

rHistAxis = new cv.HistAxes({
    channel: 2,
    bins: 256,
    ranges: [0, 1]
})

function getDescriptors(img) {
    const keypoints = SIFT.detect(img)
    const descriptors = SIFT.compute(img, keypoints)
    return descriptors
}

function cropBlackBars(image) {
    const kSize = new cv.Size(15, 15)

    const grayImage = image.bgrToGray().gaussianBlur(kSize, 10, 10).convertTo(cv.CV_8UC1, 255).threshold(1, 255, cv.THRESH_BINARY)
    const contours = grayImage.findContours(cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)
    const biggestContour = contours.sort((a, b) => {
        return b.area - a.area
    })[0]
    return image.getRegion(biggestContour.boundingRect())
}

function setSomeValuesToZeroInHist(hist) {
    hist.set(0, 0, 0)
    hist.set(1, 0, 0)
    hist.set(255, 0, 0)
    hist.set(254, 0, 0)
}

function findMatchesWithHistogram(stockArray, receivedImg) {

    const histCompMethod = cv.HISTCMP_BHATTACHARYYA
    const similarityThreshold = 0.15

    const similarStock = []

    const bHistRec = cv.calcHist(receivedImg, [bHistAxis]).convertTo(cv.CV_32F, 1 / 255)
    const gHistRec = cv.calcHist(receivedImg, [gHistAxis]).convertTo(cv.CV_32F, 1 / 255)
    const rHistRec = cv.calcHist(receivedImg, [rHistAxis]).convertTo(cv.CV_32F, 1 / 255)
    setSomeValuesToZeroInHist(bHistRec)
    setSomeValuesToZeroInHist(gHistRec)
    setSomeValuesToZeroInHist(rHistRec)

    const storedHistograms = JSON.parse(fs.readFileSync(`./histograms.txt`))

    for (let index = 0; index < stockArray.length; index++) {
        const stock = stockArray[index]
        const filePath = stock[1]

        const bHistRef = new cv.Mat(storedHistograms[filePath][0], cv.CV_32F)
        const similarityB = bHistRec.compareHist(bHistRef, histCompMethod)
        if (!(similarityB < similarityThreshold)) {
            continue
        }

        const gHistRef = new cv.Mat(storedHistograms[filePath][1], cv.CV_32F)
        const similarityG = gHistRec.compareHist(gHistRef, histCompMethod)
        if (!(similarityG < similarityThreshold)) {
            continue
        }

        const rHistRef = new cv.Mat(storedHistograms[filePath][2], cv.CV_32F)
        const similarityR = rHistRec.compareHist(rHistRef, histCompMethod)
        if (!(similarityR < similarityThreshold)) {
            continue
        }

        similarStock.push(stock.slice())
    }

    return similarStock
}

function bestMatchUsingFeatureDetector(someArray, receivedImgDescriptors) {
    let minError = 10000
    let bestMatch
    someArray.forEach(data => {
        const matches = bruteForceMatcher.match(data[0], receivedImgDescriptors)
            .sort((a, b) => {
                return a.distance - b.distance
            })
            .splice(0, 100)

        let errorSum = 0
        matches.forEach(match => {
            errorSum = match.distance + errorSum
        })

        if (errorSum < minError) {
            minError = errorSum
            bestMatch = data
        }
    })

    if (minError < 4500) {
        // console.log(`best match[4] = ${bestMatch[4]}`)
    }

    return [minError, bestMatch]
}

async function optimisedImageMatcher(imageDataArray, image) {
    image = cropBlackBars(image)

    const reducedStockDataArray = findMatchesWithHistogram(imageDataArray, image.convertTo(cv.CV_32F, 1 / 255))

    reducedStockDataArray.forEach(((stock, index) => {
        reducedStockDataArray[index][0] = JSON.parse(fs.readFileSync(stock[0]))
        reducedStockDataArray[index][0] = new cv.Mat(stock[0], cv.CV_32F)
    }))

    const receivedImgDescriptors = getDescriptors(image.convertTo(cv.CV_8U))
    const featureDetectorResult = bestMatchUsingFeatureDetector(reducedStockDataArray, receivedImgDescriptors)

    if (featureDetectorResult[0] < 9000) {
        const productFilepath = featureDetectorResult[1][1]
        const detectedImage = cv.imread(productFilepath)

        // const detectionNo = fs.readdirSync(`./Recognised Products`).length + 1
        // const detectionoDirPath = `./Recognised Products/${detectionNo}`
        // fs.mkdirSync(detectionoDirPath)
        // cv.imwrite(`${detectionoDirPath}/detected.jpeg`, detectedImage)
        // cv.imwrite(`${detectionoDirPath}/received.jpeg`, image)
        // fs.writeFileSync(`${detectionoDirPath}/min error.txt`, JSON.stringify(featureDetectorResult[0]))
    }

    return featureDetectorResult
}


function generateRandomString(length) {
    const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    let result = '';

    for (let i = 0; i < length; i++) {
        const randomIndex = Math.floor(Math.random() * characters.length);
        result += characters.charAt(randomIndex);
    }

    return result;
}


const execFileAsync = promisify(execFile);
async function optimisedImageMatcherCuda(imageDataArray, image) {

    const imgFilePath = `temp_images/${generateRandomString(16)}.jpeg`;
    console.log(imgFilePath);
    cv.imwrite(imgFilePath, image);

    const featureDetectorResult = JSON.parse((await execFileAsync('python', ['automation_modules/image_matcher.py', imgFilePath])).stdout);
    fs.unlink(imgFilePath, () => {});

    return featureDetectorResult;
}

module.exports.getProductData = getProductData;
module.exports.getSeStockData = getSeStockData;
module.exports.getStockData = getStockData;
module.exports.cvMatToArray = cvMatToArray;

module.exports.bHistAxis = bHistAxis;
module.exports.rHistAxis = rHistAxis;
module.exports.gHistAxis = gHistAxis;
module.exports.cropBlackBars = cropBlackBars;
module.exports.getDescriptors = getDescriptors;
module.exports.setSomeValuesToZeroInHist = setSomeValuesToZeroInHist;
module.exports.findMatchesWithHistogram = findMatchesWithHistogram;
module.exports.bestMatchUsingFeatureDetector = bestMatchUsingFeatureDetector;
module.exports.optimisedImageMatcher = optimisedImageMatcher;
module.exports.optimisedImageMatcherCuda = optimisedImageMatcherCuda;